### ISSUE TITLE
Native Progress Bar/tqdm for Ray

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#9233]
### Describe your feature request

Using Ray with tdqm is always really unwieldy. Either, you have to restructure your application to yield control at progress reporting points, or you are faced with a large pyramid:
```
  0%|▎                                                                                                          | 1/400 [01:37<10:48:44, 97.56s/it]
  0%|▌                                                                                                           | 2/400 [01:38<7:35:18, 68.64s/it]
  1%|▊                                                                                                           | 3/400 [01:39<5:20:10, 48.39s/it]
  1%|█                                                                                                           | 4/400 [01:41<3:46:10, 34.27s/it]
  1%|█▎                                                                                                          | 5/400 [01:42<2:40:40, 24.41s/it]
  2%|█▌                                                                                                          | 6/400 [01:43<1:54:41, 17.47s/it]
  2%|█▉                                                                                                          | 7/400 [01:45<1:22:43, 12.63s/it]
  2%|██▏                                                                                                           | 8/400 [01:46<59:54,  9.17s/it]
  2%|██▍                                                                                                           | 9/400 [01:47<44:45,  6.87s/it]
  2%|██▋                                                                                                          | 10/400 [01:49<33:55,  5.22s/it]
  3%|██▉                                                                                                          | 11/400 [01:50<25:45,  3.97s/it]
  3%|███▎                                                                                                         | 12/400 [01:52<21:26,  3.31s/it]
  3%|███▌                                                                                                         | 13/400 [01:53<17:06,  2.65s/it]
  4%|███▊                                                                                                         | 14/400 [01:54<14:31,  2.26s/it]
  4%|████                                                                                                         | 15/400 [01:55<12:43,  1.98s/it]
  4%|████▎                                                                                                        | 16/400 [01:57<11:32,  1.80s/it]
  4%|████▋                                                                                                        | 17/400 [01:58<10:14,  1.60s/it]
  4%|████▉                                                                                                        | 18/400 [01:59<10:00,  1.57s/it]
  5%|█████▏                                                                                                       | 19/400 [02:00<09:06,  1.43s/it]
  5%|█████▍                                                                                                       | 20/400 [02:02<08:46,  1.38s/it]
  5%|█████▋                                                                                                       | 21/400 [02:03<09:22,  1.48s/it]
  6%|█████▉                                                                                                       | 22/400 [02:05<08:59,  1.43s/it]
  6%|██████▎                                                                                                      | 23/400 [02:06<08:18,  1.32s/it]
  6%|██████▌                                                                                                      | 24/400 [02:07<07:59,  1.27s/it]
  6%|██████▊                                                                                                      | 25/400 [02:08<08:19,  1.33s/it]
  6%|███████                                                                                                      | 26/400 [02:10<08:50,  1.42s/it]
  7%|███████▎                                                                                                     | 27/400 [02:11<08:12,  1.32s/it]
  7%|███████▋                                                                                                     | 28/400 [02:13<08:38,  1.40s/it]
  7%|███████▉                                                                                                     | 29/400 [02:14<08:45,  1.42s/it]
  8%|████████▏                                                                                                    | 30/400 [02:15<08:00,  1.30s/it]
  8%|████████▍                                                                                                    | 31/400 [02:17<08:02,  1.31s/it]
```

One way of achieving this is by simply enabling carriage returns when Ray spits out the worker output.

Similarly, it'd be nice to enable stdout for only 1 actor at a time.

**Comment by LikvidJozsi** [comment]
This carriage return problem is created when the worker adds the pid to the start of each line printed. After doing so it calls the print function, which by default puts a newline character at the end of the string. By modifying the print calls in [print_worker_logs](https://github.com/ray-project/ray/blob/973a7a3ffa9eee83007b605cb903191d3c0ee860/python/ray/worker.py#L1033) function in worker.py by adding an extra `end=""` parameter, the issue is resolved. I have no idea if this breaks something, but it works for me nicely, if someone needs a quick&dirty fix.

**Comment by jjyao** [comment]
In general, if there are multiple tasks that are running concurrently with tqdm, the progress bar will interfere with each other if we try to fix the carriage return issue. The pyramid at least let us see the progress from each remote task clearly.

**Comment by amogkam** [comment]
Bumping priority from P3 to P2 as this is a usability improvement for libraries/Ray AIR.

**Comment by Yatagarasu50469** [comment]
For others facing this issue, here is a quick workaround/example for visualizing global progress using pool actors, though it should be adaptable for standard ray.get() calls...Hope this helps someone else.

Summary: Create an actor to hold the currently achieved progress, passing its handle, as a parameter, into calls of the intended work method. Inside the work method, issue a blocking ray.get() call to update the global actor variable. Back in the main process, use an asynchronous starmap to initialize the jobs and then regularly update a tqdm progress bar, until all results are returned. 

```
import numpy as np
import ray
import time
import warnings
from ray.util.multiprocessing import Pool
from tqdm.auto import tqdm

ray.init()

#Ray actor for holding global progress during parallel operations
@ray.remote
class Progress_Actor:
    def __init__(self): self.current = 0.0
    def update(self, amount): self.current += amount
    def getCurrent(self): return self.current

def doSomeWork(progress_Actor, amount):

    #For each the sub-tasks that need to be done
    for i in range(amount):
       
       #Do some work
        time.sleep(0.1)

        #After some work has been done, update the global actor progress using a blocking call
        _ = ray.get(progress_Actor.update.remote(1))

   #Return data; validates that all results are returned in expected order
    return amount

#Setup some variable job parameters
amounts = [10, 20, 5, 30, 25, 10]

#Initialize an actor to track how much has been done
maxProgress = float(np.sum(amounts))
pbar = tqdm(total=maxProgress, desc = 'Total', leave=True, ascii=True)
progress_Actor = Progress_Actor.remote()

#Setup jobs, passing in actor reference
futures = [(progress_Actor, amount) for amount in amounts]

#Start a pool, distribute jobs with a non-blocking async call, and lose the pool to new inputs
p = Pool(2)
results = p.starmap_async(doSomeWork, futures)
p.close()

#While all results have yet to be returned
pbar.n = 0
pbar.refresh()
while (True):

    #Update the global progress bar
    pbar.n = np.clip(round(ray.get(progress_Actor.getCurrent.remote()),2), 0, maxProgress)
    pbar.refresh()

    #Retrieve results and close the progress bar when done
    if results.ready(): 
        results = results.get()
        pbar.close()
        del progress_Actor
        break

    #Specify a polling rate to prevent excessive getCurrent() actor calls
    time.sleep(0.1)

#Ensure pool has completed
p.join()

```

**Comment by richardliaw** [comment]
addressed in #33122
